src/server/frontend_wayland/generator/emitter.cpp

```diff
@@ -0,0 +1,237 @@
+          indent{indent}
+    {
+    }
+
+    virtual ~LayoutEmitter() = default;
+
+    void emit(Emitter::State state) const override
+    {
+        state.indent += indent;
+        if (clear_line_before && !*state.on_fresh_line)
+        {
+            state.out << "\n";
+            *state.on_fresh_line = true;
+        }
+        child.emit(state);
+        if (clear_line_after && !*state.on_fresh_line)
+        {
+            state.out << "\n";
+            *state.on_fresh_line = true;
+        }
+    }
+
+private:
+    Emitter const child;
+    bool const clear_line_before;
+    bool const clear_line_after;
+    std::string const indent;
+};
+
+Emitter::Emitter(std::string const& text)
+    : impl{text.empty() ? nullptr : std::make_shared<StringEmitter>(text)}
+{
+}
+
+Emitter::Emitter(const char* text)
+    : impl{text != nullptr && *text!='\0' ?
+               std::make_shared<StringEmitter>(std::string{text}) :
+               nullptr}
+{
+}
+
+Emitter::Emitter(std::initializer_list<Emitter> const& emitters)
+    : impl{Impl::contains_valid(emitters) ?
+               std::make_shared<SeqEmitter>(emitters) :
+               nullptr}
+{
+}
+
+Emitter::Emitter(std::vector<Emitter> const& emitters)
+    : impl{Impl::contains_valid(emitters) ?
+               std::make_shared<SeqEmitter>(move(emitters)) :
+               nullptr}
+{
+}
+
+Emitter::Emitter(Line && line)
+{
+    Emitter e{line.emitters};
+    if (e.is_valid())
+    {
+        impl = std::make_shared<LayoutEmitter>(std::move(e), line.break_before, line.break_after, line.indent);
+    }
+}
+
+Emitter::Emitter(Lines && lines)
+{
+    std::vector<Emitter> l;
+    for (auto& i : lines.emitters)
+    {
+        if (i.is_valid())
+            l.push_back(Emitter{std::make_shared<LayoutEmitter>(std::move(i))});
+    }
+    if (!l.empty())
+    {
+        impl = std::make_shared<SeqEmitter>(move(l));
+    }
+}
+
+Emitter::Emitter(Block && block)
+{
+    impl = std::make_shared<LayoutEmitter>(Emitter{"{",
+                                                   Emitter{std::make_shared<LayoutEmitter>(
+                                                       Lines{block.emitters}, true, true, single_indent)},
+                                                   "}"}, true, true);
+}
+
+Emitter::Emitter(List && list)
+{
+    std::vector<Emitter> l;
+    for (auto& i : list.items)
+    {
+        if (i.is_valid())
+            l.push_back(std::move(i));
+    }
+    if (!l.empty())
+    {
+        impl = std::make_shared<SeqEmitter>(move(l), std::move(list.delimiter));
+        if (!list.indent.empty())
+            impl = std::make_shared<LayoutEmitter>(Emitter{impl}, false, false, list.indent);
+    }
+}
+
+void Emitter::emit(State state) const
+{
+    if (impl)
+    {
+        impl->emit(state);
+    }
+}
+
+Emitter::Emitter(std::shared_ptr<Impl const> impl)
+    : impl{move(impl)}
+{
+}
+
+Emitter const extern empty_line = Emitter{std::make_shared<EmptyLineEmitter>()};
```

Comment:
Huh. Yeah, I guess the `extern` specifier here is doing a thing, but I was originally taken aback by the combination of `extern` linkage *and* an initialiser.

We tend to not duplicate the `extern` specifier on the declaration (cf: `src/platform/options/default_configuration.cpp`)

Reply:
> We tend to not duplicate the extern specifier on the declaration (cf: src/platform/options/default_configuration.cpp)

I think the only reason for that is we hit a platform where it caused problems. (No, I can't remember.)